### ISSUE TITLE
GP support for TuringLang

### ISSUE THREAD

**Issue by yebai** [TuringLang/Turing.jl#777]

**Comment by cpfiffer** [comment]
Figure I should flag the JSoC person attached to this: @sshkhr. I know you're a bit busy right now, so need need to do anything right now. This is a good place to consolidate thoughts, goals, and questions on how to integrate GPs into Turing.

**Comment by sshkhr** [comment]
@cpfiffer Makes sense. I'll keep all my questions and discussion related to the GPs project in this thread for future reference

**Comment by yebai** [comment]
UPDATE: the following Julia package is also closely related, but seems no longer maintained (added to the above list):

https://github.com/adriancu/GPkit.jl/tree/master/src

**Comment by sshkhr** [comment]
@yebai and @willtebbutt apologies for the radio silence. I've been catching up on GP fundamentals and the linked implementations and got a bit overwhelmed. I'll make sure to stay in regular contact going forward. 

I had the following ideas and questions regarding the start of the coding period:

- I'll start off with the mean function implementations which are available in other libraries (mostly GPML and GaussianProcesses.jl) but not in Stheno yet. I believe that was the idea at the time of the proposal to develop Stheno and make it compatible with Turing. If this is okay I'll work on a fork of Stheno and maybe refactor the existing `means_and_kernel` module into separate `means` and `kernel` modules so as to follow closely the description given by @yebai. Would like to hear your thoughts on this @willtebbutt.

- @theogf who wrote `AugmentedGaussianProcesses.jl` has started work on `KernelFunctions.jl` to serve as an efficient kernel package to GP packages like `Stheno.jl` and `AugmentedGaussianProcesses.jl`. We can work with Théo on developing it and use it for our project.

- Flux seems to be moving away from `Tracker.jl` to `Zygote.jl`. Flux's website contains instructions on Tracker.jl usage (https://fluxml.ai/Flux.jl/stable/internals/tracker/) but Zygote.jl is still a work in progress. I'm not sure if in the future it'll be straightforward to update compatibility with `Tracker.jl` to `Zygote.jl` and would like to hear from someone more knowledgeable about automatic differentiation in Julia to help me out here regarding what would be suitable for us.

Also, I'd like to work in the form of individual PRs for each component mentioned in the project goals. I understand that this might make keeping all the updates consolidated a difficult task but I feel that getting feedback is easier this way. We can maintain a task list in this thread for more of a bird's eye view of the project status.

**Comment by willtebbutt** [comment]
> I've been catching up on GP fundamentals and the linked implementations and got a bit overwhelmed.

No problem at all :)

> If this is okay I'll work on a fork of Stheno and maybe refactor the existing means_and_kernel module into separate means and kernel modules so as to follow closely the description given by @yebai.

This is fine by me -- I can't think of any particular reason that we shouldn't do this. Please be sure to work off the `wct/refactor-composition` branch, as that has the most up-to-date implementations.

> We can work with Théo on developing it and use it for our project.

👍 

> Flux seems to be moving away from Tracker.jl to Zygote.jl. Flux's website contains instructions on Tracker.jl usage (https://fluxml.ai/Flux.jl/stable/internals/tracker/) but Zygote.jl is still a work in progress. I'm not sure if in the future it'll be straightforward to update compatibility with Tracker.jl to Zygote.jl and would like to hear from someone more knowledgeable about automatic differentiation in Julia to help me out here regarding what would be suitable for us.

This is a tricky one. On the one hand, Stheno (the refactor-composition) basically works fine with Zygote (but not Tracker), but Turing only works with Tracker (and ForwardDiff) currently. However, Tracker makes it really easy to implement custom sensitivities, so we could just implement one for `logpdf` of a GP that makes it use Zygote for just that bit of the model, and lets Tracker be used for the rest. @yebai @mohamed82008 are there any functions other than `logpdf` you can think of that we need to be able to backprop through?

> Also, I'd like to work in the form of individual PRs for each component mentioned in the project goals. I understand that this might make keeping all the updates consolidated a difficult task but I feel that getting feedback is easier this way. We can maintain a task list in this thread for more of a bird's eye view of the project status.

👍

**Comment by yebai** [comment]
>  However, Tracker makes it really easy to implement custom sensitivities, so we could just implement one for logpdf of a GP that makes it use Zygote for just that bit of the model, and lets Tracker be used for the rest. @yebai @mohamed82008 are there any functions other than logpdf you can think of that we need to be able to backprop through?

This sounds like a practical approach when moving forward -- allowing us to take advantages of both `Tracker` and `Zygote`.

**Comment by sshkhr** [comment]
> This is fine by me -- I can't think of any particular reason that we shouldn't do this. Please be sure to work off the `wct/refactor-composition` branch, as that has the most up-to-date implementations.

I'm working with this branch :+1: 

> This is a tricky one. On the one hand, Stheno (the refactor-composition) basically works fine with Zygote (but not Tracker), but Turing only works with Tracker (and ForwardDiff) currently. However, Tracker makes it really easy to implement custom sensitivities, so we could just implement one for `logpdf` of a GP that makes it use Zygote for just that bit of the model, and lets Tracker be used for the rest. @yebai @mohamed82008 are there any functions other than `logpdf` you can think of that we need to be able to backprop through?

This seems like a good workaround for now. We can look into updating the Tracker code to Zygote later once Flux releases a working update.
